optics-core/src/Optics/Internal/Optic.hs

```diff
@@ -113,12 +113,21 @@ castOptic (Optic o) = Optic (cast o)
 -- indexed, the composition preserves all the indices.
 --
 infixl 9 %
-(%) :: (Is k m, Is l m, m ~ Join k l, AppendIndices is js ks)
+(%) :: forall k l m is js ks s t u v a b. (JoinKinds k l m, AppendIndices is js ks)
     => Optic k is s t u v
     -> Optic l js u v a b
-    -> Optic m ks s t a b
-o % o' = castOptic o %% castOptic o'
-{-# INLINE (%) #-}
+      -> Optic m ks s t a b
+Optic k % Optic l = Optic m
+  where
+    km :: forall p i. Profunctor p => Optic_ m p i (Curry is i) s t u v
+    km = joinKinds @k @l @m @p k
```

Comment:
These `joinKinds` are essentially the same as `castOptic`? Should we have `castOptic`-like function which uses `joinKinds`?

Do we need `Is` at all, isn't `Is k l` `JoinKinds k l l`? I don't suggest we need to remove `Is`, I just wonder how `JoinKinds` is related.

Reply:
> These joinKinds are essentially the same as castOptic?

Yes, it's basically an extended version of `implies` you can use for both `k` and `l`.

> Should we have castOptic-like function which uses joinKinds?

`Is` is used for subtyping, while `JoinKinds` for composition. It's true that they look very similar, but have distinct use cases.

> Do we need Is at all, isn't Is k l JoinKinds k l l?

Yes, it looks like it. Functionally it could be used in place of `Is`, but it would be weird to users. Plus error messages in overlappable instances are different.

Reply:
I agree that their use cases are different, and thus repeating is not bad (especially as it is codegen).

Reply:
Is there a reason to prefer this encoding over reusing `Is` like this?
```hs
class (Is k m, Is l m) => JoinKinds k l m | k l -> m
```

Reply:
`JoinKinds` won't be `newtype` anymore. It might affect optimizer. If it doesn't, I'd like this sanity check.

Reply:
> Is there a reason to prefer this encoding over reusing `Is` like this?
> 
> ```haskell
> class (Is k m, Is l m) => JoinKinds k l m | k l -> m
> ```

We'd need to enumerate all cases anyway and we use the same codegen for it as for the `Is` instances.

I also don't know if there can be any complications like errors from `Is` being printed when overlappable error instance for `JoinKinds` is picked.

Reply:
Fair points. I'm happy with the plan here, feel free to merge.

I can imagine one might know `JoinKinds` and yet need `Is`, but only if one is doing something pretty close to `optics` internals, so let's worry about that if we ever find such a case.

In another context (`HasField`-related) I find myself wishing that one could define magically-contracted constraint synonyms, e.g. define
```hs
type Is k l = JoinKinds k l l
```
and have GHC print `Is` instead of `JoinKinds` in inferred types and error messages whenever they are equivalent. Perhaps one day...